crates/sound/src/audio_backends/pipewire.rs

```diff
@@ -1,30 +1,566 @@
+                    id: pw::spa::sys::SPA_PARAM_EnumFormat,
+                    properties: audio_info.into(),
+                }),
+            )
+            .unwrap()
+            .0
+            .into_inner();
+
+            let value_clone = values.clone();
+
+            let mut param = [Pod::from_bytes(&values).unwrap()];
+
+            let props = properties! {
+                *pw::keys::MEDIA_TYPE => "Audio",
+                *pw::keys::MEDIA_CATEGORY => "Playback",
+            };
+            // 75% of the timer period for faster updates
+            // Todo: investigate the right value to use
+            let buf_samples: u32 = info.rate * 3 / 4 / 100;
+
+            properties_set!(
+                props,
+                *pw::keys::NODE_LATENCY,
+                "{} / {}",
+                buf_samples,
+                info.rate
+            );
+            let stream = pw::stream::Stream::new(&self.core, "audio-output", props)
+                .expect("could not create new stream");
+
+            let streams = self.stream_params.clone();
+
+            let listener_stream = stream
+                .add_local_listener()
+                .state_changed(move |old, new| {
+                    debug!("State changed: {:?} -> {:?}", old, new);
+                })
+                .param_changed(move |stream, id, _data, param| {
+                    let Some(_param) = param else {
+                        return;
+                    };
+                    if id != ParamType::Format.as_raw() {
+                        return;
+                    }
+                    let mut param = [Pod::from_bytes(&value_clone).unwrap()];
+
+                    //callback to negotiate new set of streams
+                    stream
+                        .update_params(&mut param)
+                        .expect("could not update params");
+                })
+                .process(move |stream, _data| {
+                    //todo: use safe dequeue_buffer(), contribute queue_buffer()
+                    unsafe {
+                        let b: *mut pw_buffer = stream.dequeue_raw_buffer();
+                        if b.is_null() {
+                            return;
+                        }
+                        let buf = (*b).buffer;
+                        let datas = (*buf).datas;
+
+                        let p = (*datas).data as *mut u8;
+
+                        if p.is_null() {
+                            return;
+                        }
+
+                        // to calculate as sizeof(int16_t) * NR_CHANNELS
+                        let frame_size = info.channels as u32 * size_of::<u16>() as u32;
+                        let req = (*b).requested * (frame_size as u64);
+                        let mut n_bytes = cmp::min(req as u32, (*datas).maxsize);
+
+                        let stream2 = &mut streams.write().unwrap()[stream_id as usize];
+                        let req2 = stream2.buffers.front_mut().unwrap();
```

Comment:
we have to handle the case in which req2 is none because there is no new buffer. In that case, we can just fill with silence by implementing audio_pcm_info_clear_buf().

Reply:
do you mean to complete the todo in `if avail <= 0 {}` pipewire.rs:L440

Reply:
Not exactly, I am referring to the case in which `front_mut()` returns "none". We should have something like:
```rust
   let buf = match stream_reqs.buffers.front_mut() {
        Some(x) => x,
        None => return // I think in this case we have to do something similar to the case avail == 0
    };
```

Reply:
Ah I see what you mean, ok ok.